Make RELATE language-agnostic by allowing specification of Docker containers.

davis68 opened the pull request:
This contributes further towards #631, creating support for non-Python questions.

The current version (`c6943a7`) adds support for specifying Docker containers, although the default on everything is still Python.  Terminology such as `runpy` has been switched to `runcode`.  The Docker image `inducer/relate-runcode-python` has been introduced.

davis68 commented:
Looks like it's mad about the CSV import error you've fixed on another branch, but that looks like all of it.

Ah, you'll have to patch back in [this one](https://github.com/inducer/relate/pull/667/commits/62530a40881979c0194755e58da1af027772485d) or I can rebuild the PR once you've merged #667.

inducer commented:
#675 should have cleared up all the CI failures. (Sorry about having those hanging for so long.) Could you merge in master and see what happens?

inducer commented on tests/test_grades/test_csv.py:
```diff
@@ -413,8 +413,7 @@ def test_import_success_no_header(self):
         expected_file_error_msg = (
-            "Error: line contains NUL. Are you sure the file is a "
-            "CSV file other than a Microsoft Excel file?")
+            "Error: line contains NUL")
```
If you remove this change from your branch, your CI should pass.

davis68 replied:
Yeah, I ended up with a three-way merge on that block and bet on the wrong horse.  I daresay everything passes now.

BUT it's going to be pulling in the default Docker image, not the new one.  See `docker_image_run_py/Dockerfile` for details.  We should definitely test that before a merge.  It works locally but that worth about the letters it takes to type it.

davis68 commented:
BUT it's going to be pulling in the default Docker image, not the new one.  See `docker_image_run_py/Dockerfile` for details.  We should definitely test that before a merge.  It works locally but that worth about the letters it takes to type it.

inducer commented:
I've built and pushed `inducer/relate-runcode-python`. Could you apply this diff:

```diff
diff --git a/docker-image-run-py/Dockerfile b/docker-image-run-py/Dockerfile
index 4e9eacf2..f73228e1 100644
--- a/docker-image-run-py/Dockerfile
+++ b/docker-image-run-py/Dockerfile
@@ -3,9 +3,6 @@ MAINTAINER Andreas Kloeckner <inform@tiker.net>
 EXPOSE 9941
 RUN useradd runcode
 
-# Temporarily needed for pandas
-RUN echo "deb http://httpredir.debian.org/debian unstable main contrib" >> /etc/apt/sources.list
-RUN echo "deb http://httpredir.debian.org/debian experimental main contrib" >> /etc/apt/sources.list
 RUN echo 'APT::Default-Release "testing";' >> /etc/apt/apt.conf
 
 RUN apt-get update
@@ -40,8 +37,6 @@ RUN apt-get -y -o APT::Install-Recommends=0 -o APT::Install-Suggests=0 install \
 RUN apt-get clean
 RUN fc-cache
 
-RUN pip3 install numpy==1.16.1
-
 RUN mkdir -p /opt/runcode
 ADD runcode /opt/runcode/
 COPY code_feedback.py /opt/runcode/
@@ -53,7 +48,8 @@ RUN echo "savefig.dpi : 80" >> /etc/matplotlibrc
 RUN echo "savefig.bbox : tight" >> /etc/matplotlibrc
 
 # RUN pip3 install --upgrade tensorflow
+# RUN  find /usr/lib/python3/ -name '*cpython-37m*.so' -delete
+
 RUN rm -Rf /root/.cache
-RUN  find /usr/lib/python3/ -name '*cpython-37m*.so' -delete
 
 # may use ./flatten-container.sh to reduce disk space
```

(and, for future PRs (and maybe this one), could you tick the box that allows me to push?)

inducer commented:
Should this also include changing the name of the setting `RELATE_DOCKER_RUNPY_IMAGE` and pointing at the new container?

davis68 commented:
The box `Allow edits from maintainers` is checked.  What do you observe that is different with this PR than what you expect?

davis68 commented:
>Should this also include changing the name of the setting `RELATE_DOCKER_RUNPY_IMAGE` and pointing at the new container?

Yes.  My `local_settings.py` has:

```
RELATE_DOCKER_RUNPY_IMAGE = "inducer/relate-runcode-python"
```

(and will have `RELATE_DOCKER_RUNOC_IMAGE = "davis68/relate-runcode-octave"` later).

davis68 commented:
I should probably hardcode a fallback in case of an older `local_settings.py`.

inducer commented:
> `RELATE_DOCKER_RUNOC_IMAGE`

Could you spell out `RUNOCTAVE`? (I'm happy to have Python also spelled out for fairness. "OC" is a bit hard to guess otherwise...)

inducer commented:
> >Should this also include changing the name of the setting `RELATE_DOCKER_RUNPY_IMAGE` and pointing at the new container?
>
> Yes.

That's my question. Should I wait with merging this until those changes are included? I.e. could you let me know when/whether this is ready for review?

davis68 commented:
Current major references to `RELATE_DOCKER_RUNPY_IMAGE` include:

- /tests/test_pages/test_code.py has an override (see `REAL_` below)
- course/page/code.py loads it, of course

Current hardcoded references to `inducer/relate-runpy-*` include:

- tests/test_pages/utils.py refers to `REAL_RELATE_DOCKER_RUNPY_IMAGE` as `inducer/relate-runpy-i386`
- local_settings_example.py has the old version

I have just updated the Dockerfile per your diff above and also fixed `local_settings_example.py`.  If these clear, I am "reasonably confident" that things have integrated successfully, barring what the i386 version does (which I have not investigated).

@inducer My actionable question (assuming checks pass) is whether to swap out the hardcoded i386 version for the generic runcode version of the image.

davis68 commented:
@inducer :point_up: Please see the question on the hardcoded i386 Docker image.

inducer commented:
> My  actionable question (assuming checks pass) is whether to swap out the  hardcoded i386 version for the generic runcode version of the image.

Yes, please do. Sorry about the delay in responding. I've had a pretty rough two weeks, and I'm still battling a cold...

davis68 commented:
No problem.  I think I may be getting the cold going around myself...

davis68 commented:
Bumping this, then I'll finalize the Octave addition and write up documentation.

inducer commented on course/page/code.py:
```diff
@@ -443,6 +443,12 @@ class CodeQuestion(PageBaseWithTitle, PageBaseWithValue):
+
+        Optional.
+        Specific Docker image within which to run code for the participants
+        answer.
```
* Typo: participant's
* This should make brief reference to the interface that the Docker image is expected to offer. (to counter a notion that any docker image will do)
* Also state whether the image needs to be already downloaded or whether it will be downloaded on demand.

davis68 replied:
>Specific Docker image within which to run code for the participants
        answer.  This overrides the image set in the `local_settings.py`
        configuration.  The Docker image should provide two files; these are
        supplied in RELATE's standard Python Docker image by `course/page/
        code_run_backend_python.py` and `course/page/code_feedback.py`, for
        instance.  Consult `docker-image-run-py/docker-build.sh` for one
        example of a local build.  The Docker image should already be loaded
        on the system (RELATE does not pull the image automatically).

I've made these changes and added the above explanation.

inducer commented:
Sorry for the delay! Just one minor doc wrinkle to go.

inducer commented:
Thanks!
